### PR TITLE
Feature/vector predicates

### PR DESCRIPTION
**Pull request by pathtofile**
Issue https://github.com/microsoft/krabsetw/issues/122

Added 3 new predicates to make it cleaner to compare 3+ predicates, by taking in a vector of predicates.

This required the creation of a 'base' predicate class, and making all existing predicates use this.

### CODE REVIEW

File: krabs/krabs.sln [hunk 1]
```diff
@@ -78,6 +78,9 @@ EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ManagedExamples", "..\examples\ManagedExamples\ManagedExamples.csproj", "{32E71DD0-D11A-44DE-8CA8-572995AF2373}"
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "NativeExamples", "..\examples\NativeExamples\NativeExamples.vcxproj", "{D31B1A4B-8282-4AED-99FC-9AA5974B9134}"
+	ProjectSection(ProjectDependencies) = postProject
```
Comment (pathtofile) [inline krabs/krabs.sln:81]: This makes sure `NativeExamples` is always built using the latest `Microsoft.O365.Security.Native.ETW`

File: krabs/krabs/filtering/comparers.hpp [hunk 2]
```diff
@@ -30,7 +30,7 @@ namespace krabs { namespace predicates {
         struct equals
         {
             template <typename Iter1, typename Iter2>
-            bool operator()(Iter1 begin1, Iter1 end1, Iter2 begin2, Iter2 end2)
+            bool operator()(Iter1 begin1, Iter1 end1, Iter2 begin2, Iter2 end2) const
```
Comment (pathtofile) [inline krabs/krabs/filtering/comparers.hpp:33]: These are required to make the `property_*` predicates match the other predicates like `id_is`, etc.

File: krabs/krabs/filtering/predicates.hpp [hunk 3]
```diff
@@ -237,12 +249,84 @@ namespace krabs { namespace predicates {
         USHORT expected_;
     };
 
+    /**
+     * <summary>
+     *   Accepts an event if one of the predicates in the vector matches
+     * </summary>
+     */
+    struct or_filter_vector : details::predicate_base {
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply krabs/krabs/filtering/predicates.hpp]: sure, done!

File: krabs/krabs/filtering/predicates.hpp [hunk 4]
```diff
@@ -237,12 +249,84 @@ namespace krabs { namespace predicates {
...
+    private:
+        std::vector<details::predicate_base*> list_;
+    };
+
+    /**
+     * <summary>
+     *   Accepts an event if all of the predicates in the vector matches
+     * </summary>
+     */
+    struct and_filter_vector : details::predicate_base {
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply krabs/krabs/filtering/predicates.hpp]: sure, done!

**Comment by swannman** [comment]
@pathtofile this looks great! would you mind updating the version numbers in the three .nuspec files so we can push an updated nuget once we merge your change?

**Comment by pathtofile** [comment]
Updated nuspec files :-)

**Pull request merged**